### ISSUE TITLE
Matrix4f.setFrustum with zNear values < 1.0f creates distortions

### ISSUE THREAD

**Issue by farmboy0** [JOML-CI/JOML#341]
I compared JOML's implementation with https://github.com/KhronosGroup/OpenXR-SDK-Source/blob/d6b6d7a10bdcf8d4fe806b4f415fde3dd5726878/src/common/xr_linear.h#L560

and there are only 2 differences
m00 = 2.0f / (right - left) instead of (zNear + zNear) / (right - left)
m11 = 2.0f / (top - bottom) instead of (zNear + zNear) / (top - bottom)

This works without distortions even for very small zNear values like 0.05f

**Comment by farmboy0** [comment]
Thank you for pointing me to Matrixc4g.setPerspectiveOffCenterPov, somehow I missed that function and only found setFrustum.

**Comment by httpdigest** [comment]
Yeah, what can I say: Naming things and trying to find things by name is hard. :)

Actually, if you have the effective tan of the angles, instead of the angles themselves, and you don't need to compute tan(angle) yourself, then you could simply call the `setFrustum` method like `setPerspectiveOffCenterPov` does: https://github.com/JOML-CI/JOML/blob/main/src/main/java/org/joml/Matrix4f.java#L10307
```java
setFrustum(tanAngleLeft*zNear, tanAngleRight)*zNear, tanAngleDown*zNear, tanAngleUp*zNear, zNear, zFar);
```
given that you already have `tanAngleLeft`, `tanAngleRight`, `tanAngleDown` and `tanAngleUp` (without requiring _you_ to compute the tan(angle) function yourself from actual angles, of course). So, it's just about multiplying all the tan(angle) arguments with `zNear` and you have effectively the arguments that `setFrustum` needs.

**Comment by farmboy0** [comment]
I was computing the tan(angle) values myself because I get the angles from the openxr runtime, unfortunately i only found the setFrustum function to use them.
